File: dock.py
#!/usr/bin/env python3
"""herdr-pacer — dock.py

Keeps a usage pane docked along the bottom of every tab: a split pane running
`usage.sh dock`, grown to the full width of the tab.

  dock.py ensure     dock the current tab (tab and workspace hooks)
  dock.py toggle     hide the dock in every tab, or bring it back
  dock.py demo       self-check of the layout math, no herdr needed

Herdr has no pane placement for this, so it is built from stock calls, the way
herdr-sidebar builds its side column: split the bottom-left pane down, then
widen the dock by bouncing each pane beside it through a temporary tab and
splitting it back next to the pane above — pane.move keeps their processes.
Whether it is hidden lives in dock.json in $HERDR_PLUGIN_CONFIG_DIR.
"""

import fcntl
import json
import os
import socket
import sys

MARK = 'usage.sh" dock'               # how a live dock is told from other panes
LABEL = "usage"                       # the manifest pane's title
ROOT = os.path.dirname(os.path.abspath(__file__))
ROWS = 5                              # three rows of usage inside the pane border
FLOOR = 0.1                           # herdr clamps every split ratio to 0.1..0.9


# ── settings ──

def settings_path():
    base = os.environ.get("HERDR_PLUGIN_CONFIG_DIR") or os.path.join(
        os.environ.get("XDG_CONFIG_HOME") or os.path.expanduser("~/.config"),
        "herdr", "plugins", "config", "herdr-pacer")
    return os.path.join(base, "dock.json")


def hidden():
    try:
        with open(settings_path()) as f:
            return bool(json.load(f).get("hidden"))
    except (OSError, ValueError, AttributeError):
        return False


def set_hidden(value):
    path = settings_path()
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path + ".tmp", "w") as f:
        json.dump({"hidden": value}, f)
    os.replace(path + ".tmp", path)


# ── herdr socket API: one request per connection ──

class ApiError(Exception):
    pass


def call(method, **params):
    with socket.socket(socket.AF_UNIX) as s:
        s.connect(os.environ["HERDR_SOCKET_PATH"])
        s.sendall((json.dumps({"id": "herdr-pacer", "method": method,
                               "params": params}) + "\n").encode())
        buf = b""
        while not buf.endswith(b"\n"):
            chunk = s.recv(65536)
            if not chunk:
                break
            buf += chunk
    reply = json.loads(buf)
    if "error" in reply:
        raise ApiError("%s: %s" % (method, reply["error"].get("message")))
    return reply["result"]


def tree(tab_id):
    return call("layout.export", tab_id=tab_id)["layout"]


def geometry(tab_id):
    root = tree(tab_id)["root"]
    return call("pane.layout", pane_id=walk(root)[0]["pane_id"])["layout"]


# ── layout math (pure; demo() covers it) ──

def walk(node):
    return [node] if node["type"] == "pane" else walk(node["first"]) + walk(node["second"])


def is_dock(pane):
    """A live dock runs `usage.sh dock`. Herdr restores panes without their
    command, so a restored one is a shell that kept the label and our cwd."""
    command = pane.get("command")
    if command:
        return MARK in " ".join(command)
    return pane.get("label") == LABEL and pane.get("cwd") == ROOT


def bottom_left(panes, area):
    for pane in panes:
        r = pane["rect"]
        if r["x"] == 0 and r["y"] + r["height"] == area["height"]:
            return pane["pane_id"]
    return None


def full_width(panes, area, dock_id):
    return any(p["pane_id"] == dock_id and p["rect"]["width"] >= area["width"] for p in panes)


def widen_step(panes, area, dock_id):
    """(pane beside the dock, pane above the dock), or None once the dock
    spans the tab. Moving the first next to the second widens the dock."""
    me = next((p["rect"] for p in panes if p["pane_id"] == dock_id), None)
    if me is None or me["width"] >= area["width"]:
        return None
    beside = above = None
    for pane in panes:
        r = pane["rect"]
        if pane["pane_id"] == dock_id:
            continue
        if r["x"] == me["x"] + me["width"] and r["y"] <= me["y"] < r["y"] + r["height"]:
            beside = pane["pane_id"]
        if r["x"] == me["x"] and r["y"] + r["height"] == me["y"]:
            above = pane["pane_id"]
    return (beside, above) if beside and above else None


def dock_ratio(height):
    """The share the panes above keep, so the dock gets ROWS rows."""
    return 1 - min(max(ROWS / max(height, 1), FLOOR), 1 - FLOOR)


# ── actions ──

def tabs():
    return call("tab.list")["tabs"]


def focused_tab():
    return next((t["tab_id"] for t in tabs() if t.get("focused")), None)


def event_tab():
    try:
        event = json.loads(os.environ.get("HERDR_PLUGIN_EVENT_JSON") or "{}")
    except ValueError:
        event = {}
    data = event.get("data", event)
    return data.get("tab_id") or (data.get("tab") or {}).get("tab_id") or \
        os.environ.get("HERDR_TAB_ID") or focused_tab()


def close_docks(tab_id, keep=None):
    for pane in walk(tree(tab_id)["root"]):
        if is_dock(pane) and pane["pane_id"] != keep:
            call("pane.close", pane_id=pane["pane_id"])


def open_dock(tab_id):
    layout = geometry(tab_id)
    target = bottom_left(layout["panes"], layout["area"])
    if not target:
        return
    opened = call("plugin.pane.open", plugin_id="herdr-pacer", entrypoint="dock",
                  placement="split", target_pane_id=target, direction="down", focus=False)
    dock_id = opened["plugin_pane"]["pane"]["pane_id"]
    for _ in range(8):
        layout = geometry(tab_id)
        step = widen_step(layout["panes"], layout["area"], dock_id)
        if not step:
            break
        beside, above = step
        bounced = call("pane.move", pane_id=beside, destination={"type": "new_tab"}, focus=False)
        call("pane.move", pane_id=bounced["move_result"]["pane"]["pane_id"], focus=False,
             destination={"type": "tab", "tab_id": tab_id, "target_pane_id": above,
                          "split": "right"})
    root = tree(tab_id)["root"]
    if root["type"] == "split" and root["direction"] == "down" and is_dock(root["second"]):
        call("layout.set_split_ratio", tab_id=tab_id, path=[],
             ratio=dock_ratio(geometry(tab_id)["area"]["height"]))


def ensure(tab_id):
    """Adds a dock when the tab has none along its bottom. One already there is
    left alone, so a size dragged by hand stays and nothing redraws."""
    try:
        layout = tree(tab_id)
    except ApiError:
        return                                  # gone already, e.g. a bounce tab
    if layout.get("zoomed"):
        return
    panes = walk(layout["root"])
    found = [p["pane_id"] for p in panes if is_dock(p)]
    if found and len(panes) == 1:
        return                                  # the dock is the tab's only pane
    rects = geometry(tab_id)
    live = {p["pane_id"] for p in panes if p.get("command")}
    keep = next((d for d in found if d in live and
                 full_width(rects["panes"], rects["area"], d)), None)
    if len(found) > (keep is not None):
        close_docks(tab_id, keep)               # restored shells, duplicates
    if keep is None:
        focused = layout["focused_pane_id"]
        open_dock(tab_id)
        if tree(tab_id)["focused_pane_id"] != focused:
            call("pane.focus", pane_id=focused)


def toggle():
    set_hidden(not hidden())
    if not hidden():
        ensure(focused_tab())
        return
    # the tab we run from goes last: its dock may be the caller (the ✕ button),
    # and closing that pane ends this process
    here = os.environ.get("HERDR_TAB_ID")
    for tab in sorted(tabs(), key=lambda t: t["tab_id"] == here):
        close_docks(tab["tab_id"])


def locked(blocking):
    state = os.environ.get("HERDR_PLUGIN_STATE_DIR") or os.path.join(
        os.environ.get("XDG_STATE_HOME") or os.path.expanduser("~/.local/state"), "herdr-pacer")
    os.makedirs(state, exist_ok=True)
    handle = open(os.path.join(state, "dock.lock"), "w")
    try:
        fcntl.flock(handle, fcntl.LOCK_EX | (0 if blocking else fcntl.LOCK_NB))
    except BlockingIOError:
        return None                             # focus events come in bursts
    return handle


def main(argv):
    command = argv[1] if len(argv) > 1 else "ensure"
    if command == "demo":
        return demo()
    if command not in ("ensure", "toggle"):
        print(__doc__, file=sys.stderr)
        return 2
    lock = locked(blocking=command == "toggle")
    if lock is None:
        return 0
    with lock:
        if command == "toggle":
            toggle()
        elif not hidden():
            ensure(event_tab())
    return 0


def demo():
    area = {"x": 0, "y": 0, "width": 100, "height": 40}
    left = {"pane_id": "l", "rect": {"x": 0, "y": 0, "width": 50, "height": 35}}
    dock = {"pane_id": "d", "rect": {"x": 0, "y": 35, "width": 50, "height": 5}}
    right = {"pane_id": "r", "rect": {"x": 50, "y": 0, "width": 50, "height": 40}}
    assert bottom_left([right, left, dock], area) == "d"
    assert bottom_left([right, left], area) is None
    assert widen_step([left, dock, right], area, "d") == ("r", "l")
    wide = dict(dock, rect=dict(dock["rect"], width=100))
    assert widen_step([left, wide], area, "d") is None and full_width([wide], area, "d")
    assert abs(dock_ratio(40) - 0.875) < 1e-9 and abs(dock_ratio(80) - 0.9) < 1e-9
    assert is_dock({"pane_id": "d", "command": ["sh", "-c", 'exec "x/usage.sh" dock']})
    restored = {"pane_id": "x", "label": LABEL, "cwd": ROOT}
    assert is_dock(restored) and not is_dock(dict(restored, cwd="/home"))
    assert not is_dock({"pane_id": "y", "label": LABEL, "command": ["vim"]})
    print("dock.py demo: ok")
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main(sys.argv))
    except (ApiError, OSError, KeyError) as err:
        print("herdr-pacer dock: %s" % err, file=sys.stderr)
        sys.exit(1)

File: src/dock.rs
//! Keeps a usage pane docked along the bottom of every tab.
//!
//! Herdr has no pane placement for this, so it is built from stock calls, the
//! way herdr-sidebar builds its side column: split the bottom-left pane down,
//! then widen the dock by bouncing each pane beside it through a temporary tab
//! and splitting it back next to the pane above — pane.move keeps their
//! processes. Whether it is hidden lives in settings.json.

use crate::herdr::{self, call, Result};
use crate::settings;
use serde_json::{json, Value};
use std::fs::File;
use std::path::PathBuf;

const MARK: &str = "herdr-pacer\" dock"; // how a live dock is told from other panes
const LABEL: &str = "usage"; // the manifest pane's title
const FLOOR: f64 = 0.1; // herdr clamps every split ratio to 0.1..0.9

fn hidden() -> bool {
    settings::load().hidden
}

fn set_hidden(value: bool) {
    settings::save(&settings::Settings { hidden: value, ..settings::load() });
}

/// The dock's height: its border, a title row, and a row per window shown.
fn rows() -> f64 {
    let windows = settings::load().windows.iter().filter(|&&on| on).count().max(1);
    (3 + windows) as f64
}

/// The plugin checkout: $HERDR_PLUGIN_ROOT, or three levels above
/// target/release/herdr-pacer.
pub fn root() -> PathBuf {
    std::env::var_os("HERDR_PLUGIN_ROOT").map(PathBuf::from).unwrap_or_else(|| {
        let exe = std::env::current_exe().ok().and_then(|p| p.canonicalize().ok()).unwrap_or_default();
        exe.ancestors().nth(3).map(PathBuf::from).unwrap_or_default()
    })
}

// ── layout math (pure; the tests cover it) ──

#[derive(Clone, Copy, Debug, PartialEq)]
struct Rect {
    x: i64,
    y: i64,
    w: i64,
    h: i64,
}

fn rect(v: &Value) -> Rect {
    let n = |k: &str| v[k].as_i64().unwrap_or(0);
    Rect { x: n("x"), y: n("y"), w: n("width"), h: n("height") }
}

fn walk(node: &Value) -> Vec<&Value> {
    match node["type"].as_str() {
        Some("pane") => vec![node],
        _ => [walk(&node["first"]), walk(&node["second"])].concat(),
    }
}

/// A live dock runs `herdr-pacer dock`. Herdr restores panes without their
/// command, so a restored one is a shell that kept the label and our cwd.
fn is_dock(pane: &Value, root: &str) -> bool {
    match pane["command"].as_array() {
        Some(argv) => {
            let line = argv.iter().filter_map(Value::as_str).collect::<Vec<_>>().join(" ");
            line.contains(MARK) || line.contains("usage.sh\" dock") // a dock from before the port
        }
        None => pane["label"] == LABEL && pane["cwd"] == root,
    }
}

fn bottom_left(panes: &[(String, Rect)], area: Rect) -> Option<String> {
    panes.iter().find(|(_, r)| r.x == 0 && r.y + r.h == area.h).map(|(id, _)| id.clone())
}

fn full_width(panes: &[(String, Rect)], area: Rect, dock: &str) -> bool {
    panes.iter().any(|(id, r)| id == dock && r.w >= area.w)
}

/// (pane beside the dock, pane above the dock), or None once the dock spans
/// the tab. Moving the first next to the second widens the dock.
fn widen_step(panes: &[(String, Rect)], area: Rect, dock: &str) -> Option<(String, String)> {
    let me = panes.iter().find(|(id, _)| id == dock)?.1;
    if me.w >= area.w {
        return None;
    }
    let beside = panes.iter().find(|(id, r)| id != dock && r.x == me.x + me.w && r.y <= me.y && me.y < r.y + r.h)?;
    let above = panes.iter().find(|(id, r)| id != dock && r.x == me.x && r.y + r.h == me.y)?;
    Some((beside.0.clone(), above.0.clone()))
}

/// The share the panes above keep, so the dock gets `rows` rows.
fn dock_ratio(rows: f64, height: i64) -> f64 {
    1.0 - (rows / height.max(1) as f64).clamp(FLOOR, 1.0 - FLOOR)
}

// ── herdr ──

fn tree(tab: &str) -> Result<Value> {
    Ok(call("layout.export", json!({ "tab_id": tab }))?["layout"].take())
}

fn geometry(tab: &str) -> Result<(Vec<(String, Rect)>, Rect)> {
    let layout = tree(tab)?;
    let first = walk(&layout["root"]).first().and_then(|p| p["pane_id"].as_str()).unwrap_or_default().to_string();
    let g = call("pane.layout", json!({ "pane_id": first }))?;
    let panes = g["layout"]["panes"]
        .as_array()
        .into_iter()
        .flatten()
        .map(|p| (p["pane_id"].as_str().unwrap_or_default().to_string(), rect(&p["rect"])))
        .collect();
    Ok((panes, rect(&g["layout"]["area"])))
}

fn tabs() -> Result<Vec<Value>> {
    Ok(call("tab.list", json!({}))?["tabs"].as_array().cloned().unwrap_or_default())
}

fn focused_tab() -> Result<Option<String>> {
    Ok(tabs()?.iter().find(|t| t["focused"] == true).and_then(|t| t["tab_id"].as_str()).map(String::from))
}

fn event_tab() -> Result<Option<String>> {
    let event: Value = std::env::var("HERDR_PLUGIN_EVENT_JSON")
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap_or(Value::Null);
    let data = if event["data"].is_object() { &event["data"] } else { &event };
    let tab = data["tab_id"].as_str().or(data["tab"]["tab_id"].as_str()).map(String::from);
    match tab.or_else(|| std::env::var("HERDR_TAB_ID").ok()) {
        Some(tab) => Ok(Some(tab)),
        None => focused_tab(),
    }
}

fn close_docks(tab: &str, keep: Option<&str>) -> Result<()> {
    let root = root().to_string_lossy().into_owned();
    let layout = tree(tab)?;
    for pane in walk(&layout["root"]) {
        let id = pane["pane_id"].as_str().unwrap_or_default();
        if is_dock(pane, &root) && Some(id) != keep {
            call("pane.close", json!({ "pane_id": id }))?;
        }
    }
    Ok(())
}

fn open_dock(tab: &str) -> Result<()> {
    let (panes, area) = geometry(tab)?;
    let Some(target) = bottom_left(&panes, area) else { return Ok(()) };
    let opened = call(
        "plugin.pane.open",
        json!({ "plugin_id": "herdr-pacer", "entrypoint": "dock", "placement": "split",
                "target_pane_id": target, "direction": "down", "focus": false }),
    )?;
    let dock = opened["plugin_pane"]["pane"]["pane_id"].as_str().unwrap_or_default().to_string();
    for _ in 0..8 {
        let (panes, area) = geometry(tab)?;
        let Some((beside, above)) = widen_step(&panes, area, &dock) else { break };
        let bounced = call("pane.move", json!({ "pane_id": beside, "destination": { "type": "new_tab" }, "focus": false }))?;
        call(
            "pane.move",
            json!({ "pane_id": bounced["move_result"]["pane"]["pane_id"], "focus": false,
                    "destination": { "type": "tab", "tab_id": tab, "target_pane_id": above, "split": "right" } }),
        )?;
    }
    size(tab)
}

/// Gives the dock its height, when it spans the tab as the root split's bottom.
fn size(tab: &str) -> Result<()> {
    let root_node = tree(tab)?["root"].take();
    let own_root = root().to_string_lossy().into_owned();
    if root_node["type"] == "split" && root_node["direction"] == "down" && is_dock(&root_node["second"], &own_root) {
        let (_, area) = geometry(tab)?;
        call("layout.set_split_ratio", json!({ "tab_id": tab, "path": [], "ratio": dock_ratio(rows(), area.h) }))?;
    }
    Ok(())
}

/// Resizes every dock to the windows now shown, after a settings change.
pub fn resize_all() -> Result<()> {
    let _lock = lock(true);
    for tab in tabs()? {
        let _ = size(tab["tab_id"].as_str().unwrap_or_default());
    }
    Ok(())
}

/// Adds a dock when the tab has none along its bottom. One already there is
/// left alone, so a size dragged by hand stays and nothing redraws.
fn ensure(tab: &str) -> Result<()> {
    let Ok(layout) = tree(tab) else {
        return Ok(()); // gone already, e.g. a bounce tab
    };
    if layout["zoomed"] == true {
        return Ok(());
    }
    let root = root().to_string_lossy().into_owned();
    let panes = walk(&layout["root"]);
    let found: Vec<&str> = panes.iter().filter(|p| is_dock(p, &root)).filter_map(|p| p["pane_id"].as_str()).collect();
    if !found.is_empty() && panes.len() == 1 {
        return Ok(()); // the dock is the tab's only pane
    }
    let (rects, area) = geometry(tab)?;
    let live = |id: &str| panes.iter().any(|p| p["pane_id"] == id && p["command"].is_array());
    let keep = found.iter().copied().find(|d| live(d) && full_width(&rects, area, d));
    if found.len() > keep.is_some() as usize {
        close_docks(tab, keep)?; // restored shells, duplicates
    }
    if keep.is_none() {
        let focused = layout["focused_pane_id"].as_str().unwrap_or_default().to_string();
        open_dock(tab)?;
        if tree(tab)?["focused_pane_id"] != focused.as_str() {
            call("pane.focus", json!({ "pane_id": focused }))?;
        }
    }
    Ok(())
}

/// Hides the dock in every tab, or brings it back in the current one (the
/// others dock as they are visited).
pub fn toggle() -> Result<()> {
    let _lock = lock(true);
    set_hidden(!hidden());
    if !hidden() {
        return focused_tab()?.map_or(Ok(()), |tab| ensure(&tab));
    }
    // the tab we run from goes last: its dock may be the caller (the ✕ button),
    // and closing that pane ends this process
    let here = std::env::var("HERDR_TAB_ID").ok();
    let mut all = tabs()?;
    all.sort_by_key(|t| t["tab_id"].as_str() == here.as_deref());
    for tab in all {
        close_docks(tab["tab_id"].as_str().unwrap_or_default(), None)?;
    }
    Ok(())
}

/// Closes every dock and docks the current tab again, so an upgrade replaces
/// docks still running an older build. False when the dock is hidden.
pub fn restart() -> Result<bool> {
    let _lock = lock(true);
    for tab in tabs()? {
        close_docks(tab["tab_id"].as_str().unwrap_or_default(), None)?;
    }
    if hidden() {
        return Ok(false);
    }
    focused_tab()?.map_or(Ok(()), |tab| ensure(&tab))?;
    Ok(true)
}

/// The tab and workspace hooks. Focus events come in bursts, so a busy lock
/// means another copy is already on it.
pub fn ensure_event() -> Result<()> {
    let Some(_lock) = lock(false) else { return Ok(()) };
    if hidden() {
        return Ok(());
    }
    event_tab()?.map_or(Ok(()), |tab| ensure(&tab))
}

fn lock(blocking: bool) -> Option<File> {
    let file = File::create(crate::usage::state_dir().join("dock.lock")).ok()?;
    match blocking {
        true => file.lock().ok()?,
        false => file.try_lock().ok()?,
    }
    Some(file)
}

pub fn run(command: &str) -> herdr::Result<()> {
    match command {
        "ensure" => ensure_event(),
        "toggle" => toggle(),
        _ => unreachable!(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn pane(id: &str, x: i64, y: i64, w: i64, h: i64) -> (String, Rect) {
        (id.into(), Rect { x, y, w, h })
    }

    #[test]
    fn widening_a_bottom_dock() {
        let area = Rect { x: 0, y: 0, w: 100, h: 40 };
        let (left, dock, right) = (pane("l", 0, 0, 50, 35), pane("d", 0, 35, 50, 5), pane("r", 50, 0, 50, 40));
        assert_eq!(bottom_left(&[right.clone(), left.clone(), dock.clone()], area).as_deref(), Some("d"));
        assert_eq!(bottom_left(&[right.clone(), left.clone()], area), None);
        assert_eq!(widen_step(&[left.clone(), dock, right], area, "d"), Some(("r".into(), "l".into())));
        let wide = pane("d", 0, 35, 100, 5);
        assert_eq!(widen_step(&[left, wide.clone()], area, "d"), None);
        assert!(full_width(&[wide], area, "d"));
        assert!((dock_ratio(5.0, 40) - 0.875).abs() < 1e-9 && (dock_ratio(5.0, 80) - 0.9).abs() < 1e-9);
    }

    #[test]
    fn telling_docks_apart() {
        let live = json!({ "command": ["sh", "-c", "exec \"/p/target/release/herdr-pacer\" dock"] });
        let old = json!({ "command": ["sh", "-c", "exec \"/p/usage.sh\" dock"] });
        let restored = json!({ "label": LABEL, "cwd": "/p" });
        assert!(is_dock(&live, "/p") && is_dock(&old, "/p") && is_dock(&restored, "/p"));
        assert!(!is_dock(&json!({ "label": LABEL, "cwd": "/home" }), "/p"));
        assert!(!is_dock(&json!({ "label": LABEL, "command": ["vim"] }), "/p"));
    }
}
